lib/container.ts

```diff
@@ -79,310 +70,281 @@ export class Container extends (EventEmitter as {
+				const stdout = new Stream.PassThrough();
+				const stderr = new Stream.PassThrough();
 				stream.on('error', reject);
 				stream.on('end', async () => {
-					stdoutStream.end();
-					stderrStream.end();
-					const inspect = await exec.inspect();
-					resolve({
-						stdout: Buffer.concat(stdout).toString(),
-						stderr: Buffer.concat(stderr).toString(),
-						exitCode: inspect.ExitCode,
-					});
-				});
-				this.docker.modem.demuxStream(stream, stdoutStream, stderrStream);
-				stderrStream.on('data', d => {
-					stderr.push(d);
+					stdout.end();
+					stderr.end();
 				});
-				stdoutStream.on('data', d => {
-					stdout.push(d);
+				this.docker.modem.demuxStream(stream, stdout, stderr);
+
+				resolve({
+					exec,
+					stderr,
+					stdout,
 				});
 			});
 		});
 	}
 
-	private async executeCommand(command: string[]): Promise<number> {
-		this.emit('commandExecute', command.join(' '));
-		// first create an exec instance
+	private async runActionGroupCommand(command: string): Promise<number> {
+		this.emit('commandExecute', command);
+		const dockerCommand = Container.generateContainerCommand(command);
+		const exe = await this.executeCommand(dockerCommand);
+		let exitCode: number;
+		await waitForCommandCompletion(exe, ev => {
+			switch (ev.type) {
+				case CommandExecutionArtifactType.EXIT_CODE:
+					exitCode = ev.code;
+					this.emit('commandReturn', exitCode);
+					break;
+				case CommandExecutionArtifactType.STDERR_DATA:
+				case CommandExecutionArtifactType.STDOUT_DATA:
+					this.emit('commandOutput', {
+						data: ev.data,
+						isStderr: ev.type === CommandExecutionArtifactType.STDERR_DATA,
+					});
+					break;
+			}
+		});
+		return exitCode!;
+	}
+
+	private async performStagedCopy(
+		stage: StageDependentActionGroup,
+		containers: StageContainers,
+	): Promise<void> {
+		// First we need to request the tar stream from the container the source resides in
+		const container = containers[stage.stageDependency];
+		if (container == null) {
+			throw new InternalInconsistencyError(
+				`Attempt to copy from stage without a container given. StageIdx: ${
+					stage.stageDependency
+				}`,
+			);
+		}
+
+		for (const copy of stage.copies) {
+			await copyToStage(container, this, copy);
+		}
+	}
+
+	private async addFiles(operations: LocalAddOperation[]): Promise<void> {
+		const pack = tar.pack();
+		await Bluebird.each(operations, operation => {
+			return addFileToTarPack(
+				pack,
+				Path.join(this.buildContext, operation.fromPath),
+				operation.toPath,
+			);
+		});
+
+		pack.finalize();
+		await this.addFilesToContainer(pack, '/');
+	}
+
+	private async deleteFiles(files: string[]): Promise<void> {
+		// TODO: There's currently no way to delete an entire
+		// directory (nor the interface to tell livepush that
+		// a directory has been deleted). This could cause a
+		// divergence in expected and actual runtime, so it
+		// should be handled
+		await Bluebird.map(files, async f => {
```

Comment:
Do we need to delete all the files in parallel?

Reply:
We don't need to, but the aim of the game is speed so it makes sense. I did some testing and it doesn't seem to cause that big an issue to the docker daemon. If it does become an issue, I'll probably set an upper limit on the amount of promises that can be started.